l1-contracts/src/UniFiAVSManager.sol

```diff
@@ -212,6 +233,153 @@ contract UniFiAVSManager is UniFiAVSManagerStorage, IUniFiAVSManager, UUPSUpgrad
+                blsPubKeyHash: blsPubKeyHash,
+                validatorIndex: validators[i].index
+            });
+
+            newValidatorCount++;
+        }
+
+        // Update the operator's validator count
+        OperatorData storage operator = $.operators[msg.sender];
+        operator.validatorCount += uint128(newValidatorCount);
+    }
+
+    /**
+     * @inheritdoc IUniFiAVSManager
+     * @dev Restricted in this context is like `whenNotPaused` modifier from Pausable.sol
+     */
+    function verifyValidatorSignatures(bytes32[] calldata blsPubKeyHashes) external restricted {
+        UniFiAVSStorage storage $ = _getUniFiAVSManagerStorage();
+
+        for (uint256 i = 0; i < blsPubKeyHashes.length; i++) {
+            bytes32 blsPubKeyHash = blsPubKeyHashes[i];
+            ValidatorData storage validator = $.validators[blsPubKeyHash];
+            address operator = validator.operator;
+            ValidatorRegistrationData memory validatorRegistrationData = $.validatorRegistrations[blsPubKeyHash];
+
+            // Calculate the hash using EIP-712
+            BN254.G1Point memory messageHash = blsMessageHash({
+                operator: operator,
+                salt: validatorRegistrationData.salt,
+                expiry: validatorRegistrationData.expiry,
+                index: validator.index
+            });
+
+            // Use the stored signature for verification
+            bool isValid = BLSSignatureCheckerLib.isBlsSignatureValid(
+                validatorRegistrationData.pubkeyG1,
+                validatorRegistrationData.pubkeyG2,
+                validatorRegistrationData.registrationSignature,
+                messageHash
+            );
+
+            if (!isValid) {
+                _slashAndDeregisterValidator(blsPubKeyHash);
+            }
+        }
+    }
+
+    /**
+     * @inheritdoc IUniFiAVSManager
+     * @dev Restricted in this context is like `whenNotPaused` modifier from Pausable.sol
+     */
+    function slashValidatorsWithInvalidIndex(BeaconChainHelperLib.InclusionProof[] calldata proofs)
+        external
+        restricted
+    {
+        for (uint256 i = 0; i < proofs.length; i++) {
+            BeaconChainHelperLib.InclusionProof memory proof = proofs[i];
+            bytes32 blsPubKeyHash = proof.validator[0];
+            (uint64 index,) = _validateProofAndGetValidatorKeys(proof, false);
+
+            if (index != proof.validatorIndex) {
+                _slashAndDeregisterValidator(blsPubKeyHash);
+            }
+        }
+    }
+
+    /**
+     * @inheritdoc IUniFiAVSManager
+     * @dev Restricted in this context is like `whenNotPaused` modifier from Pausable.sol
+     */
+    function slashValidatorsWithInvalidPubkey(BeaconChainHelperLib.InclusionProof[] calldata proofs)
+        external
+        restricted
+    {
+        for (uint256 i = 0; i < proofs.length; i++) {
+            BeaconChainHelperLib.InclusionProof memory proof = proofs[i];
```

Comment:
You can delete this line to avoid copying calldata to memory (gas optimisation), just use `proofs[i]` instead of `proof` in the rest of the function 

`_validateProofAndGetValidatorKeys` can be refactored to work with calldata as well.